R/data.table.R

```diff
@@ -1892,72 +1890,293 @@ as.matrix.data.table = function(x, rownames=NULL, rownames.value=NULL, ...) {
+                                                     
+    which.not.recursive = sort(c(which.not.recursive, which.maybe.is.recursive))
+    
+    # Coerce to list as necessary
+    for (j in which.not.recursive) {
+      X[[j]] = as.list(X[[j]])
+    }
+    
+    X.info = column_properties(X, X.info, cols.to.update = which.not.recursive)
+  }
+  
+  # Other classes from suggested packages that have special type conversion rules
+  if (!any.non.atomic && any_col_is(X.info$uniq.classes, "integer64") && length(X.info$uniq.classes) > 1L) {
+    # Determine target class we're converting to 
+    target.class = NULL
+    if (!requireNamespace("bit64", quietly=TRUE)) 
+       stop("coercion to or from integer64 columns requires the bit64 package") # nocov
+    if (all_col_is(X.info$uniq.classes, c("integer64", "logical", "integer", "raw")))
+      target.class = "integer64"
+    else if (all_col_is(X.info$uniq.classes, c("integer64", "numeric", "complex", "character")))
+      target.class = "character" # integer64 cannot reliably be converted to numeric or complex
+    
+    if (!is.null(target.class)) { # if NULL fallback on type coercion
+      # no method as.integer64.raw so we need to do two-step conversion
+      if (target.class == "integer64" && any_col_is(X.info$uniq.classes, "raw")) {
+        which.convert = which(col_is(X.info$classes, "raw"))
+        for (j in which.convert) {
+          X[[j]] = as.integer(X[[j]])
+        }
+        X.info = column_properties(X, X.info, cols.to.update = which.convert)
+      }
+      
+      # Which columns need to be coerced?
+      which.convert = which(!col_is(X.info$classes, target.class))
+
+      # Coerce the columns
+      for (j in which.convert) {
+        switch(target.class,
+               "integer64"={ X[[j]] = bit64::as.integer64(X[[j]]) },
+               "character"={ X[[j]] = as.character(X[[j]]) })
+      }
+      
+      X.info = column_properties(X, X.info,  cols.to.update = which.convert) # update column properties info
+    }
+  }
+  
+  # If columns still do not all have the same type we need to coerce
+  if (!any.non.atomic && length(X.info$uniq.types) > 1L) { 
+    type.order = c("raw"=1L, "logical"=2L, "integer"=3L, "double"=4L, 
+                   "complex"=5L, "character"=6L, "list"=7L)
+    target.type = names(which.max(type.order[X.info$uniq.types]))
+    which.convert = which(!col_is(X.info$types, target.type))
+    
+    # Coerce the columns
+    for (j in which.convert) {
+      switch(target.type,
+             # raw impossible to reach - no coercion necessary if all columns are raw
+             "logical"={ X[[j]] = as.logical(X[[j]]) },
+             "integer"={ X[[j]] = as.integer(X[[j]]) },
+             "double"={ X[[j]] = as.double(X[[j]]) },
+             "complex"={ X[[j]] = as.complex(X[[j]]) },
+             "character"={ X[[j]] = as.character(X[[j]]) }
```

Comment:
It seems quite wasteful to allocate memory for a new column whenever a column has to be coerced to a different type/class, only to then have to copy that memory again into the new matrix in the C code. For the atomic types at least, we may just be able to set a target type, pass that to the C code, then do the coercion there at copy-time. This would slow down the copy loop in the C code as you'd have to check and coerce each column - one way around this would be to pass the vectors of unique column types and individual column types to write a nested for loop iterating along columns of each type separately to avoid slowdowns due to branch prediction. Perhaps some memory profiling would be useful here to see if this is worth doing.

Reply:
`assign.c:memrecycle` does that; new feature 29 in 1.12.4 refers to zero-copy type coercion.